cnap/core/rtdb.py

```diff
@@ -0,0 +1,154 @@
+"""
+Runtime Database (RTDB) is designed to cache the runtime information and
+use an in-memory database service (like Redis,Kafka) as the backend
+service.
+An abstract class `RuntimeDatabaseBase` is defined as common abstract
+methods to manage the tables, and class `RedisDB` is an implementation
+for Redis backend service.
+"""
+
+import json
+import logging
+
+from abc import ABC, abstractmethod
+
+import redis
+
+LOG = logging.getLogger(__name__)
+
+class RuntimeDatabaseBase(ABC):
+    """
+    Abstract base class to manage runtime database tables.
+    """
+
+    @abstractmethod
+    def connect(self) -> bool:
+        """
+        Connect to the runtime database
+        """
+        raise NotImplementedError("Subclasses should implement this")
+
+    @abstractmethod
+    def save_table_object_dict(self, table: str, obj: str, d: dict) -> None:
+        """
+        Save a dict value for an object in a table
+        """
+
+    @abstractmethod
+    def get_table_object_dict(self, table: str, obj: str) -> dict:
+        """
+        Get a dict value for an object from a table
+        """
+
+    @abstractmethod
+    def get_all_table_objects_dict(self, table: str) -> dict:
+        """
+        Get all dict values from a table
+        """
+
+    @abstractmethod
+    def check_table_object_exist(self, table: str, obj: str) -> bool:
+        """
+        Check whether given object exists in given table
+        """
+
+    @abstractmethod
+    def del_table_object(self, table: str, obj: str) -> None:
+        """
+        Delete an object from a given table
+        """
+
+class RedisDB(RuntimeDatabaseBase):
+
+    """
+    Redis backend implementation for runtime database
+    """
+
+    def __init__(self):
+        self._conn = None
+
+    def connect(self, host: str = "127.0.0.1", port: int = 6379, db: int = 0) -> bool:
+        """
+        Connect to Redis server.
+        Args:
+            host: Redis server host hostname/IP
+            port: Redis server port
+            db: Database number in Redis server
+        Returns:
+            bool: Indicate whether connecting to the Redis server successfully
+        """
+        self._conn = redis.Redis(host=host, port=port, db=db)
+        try:
+            self._conn.ping()
+        except redis.exceptions.ConnectionError:
+            LOG.error("Failed to connect to Redis")
+            return False
+        return True
+
+    def save_table_object_dict(self, table: str, obj: str, d: dict) -> None:
+        """
+        Save a dict value for an object in a table.
+        Args:
+            table: The name of the table
+            obj: The name of the object
+            d: The value to be saved
+        Returns:
+            None
+        """
+        LOG.debug("[Redis] Save => table: %s, obj: %s", table, obj)
+        self._conn.hset(table, obj, json.dumps(d))
```

Comment:
Maybe we should check if d exists or not, to avoid json dumping on none or invalid object?

Reply:
> Maybe we should check if d exists or not, to avoid json dumping on none or invalid object?

Yes, I can add some checking for this. So we don't have a scenario of setting an object to none, right?

Reply:
I am not sure if it exists a scenario, we create a empty object, but it seems make sense.
Maybe, if d is null/none, we should store a empty dict to its object?

Reply:
> I am not sure if it exists a scenario, we create a empty object, but it seems make sense. Maybe, if d is null/none, we should store a empty dict to its object?

How about this:
- For `obj`, I think we should check it, it should not be null/none.
- For `d`, I think we can ignore this checking. If d is null/none, json.dumps(d) will return an empty dict.

Reply:
It makes sense.